### PR TITLE
Added async_send_request_return_request to return the originating request

### PR DESCRIPTION
**Pull request by esteve**
Connects to #35

### CODE REVIEW

**Comment by esteve** [comment]
I'm not sure about the name of the method, so I'd appreciate if anyone comes with a better one :-)

**Comment by esteve** [comment]
@tfoote let me know if this is what you had in mind or if it needs to be reworked.

**Comment by esteve** [comment]
CI jobs:

http://ci.ros2.org/job/ros2_batch_ci_linux/604/
http://ci.ros2.org/job/ros2_batch_ci_osx/490/
http://ci.ros2.org/job/ros2_batch_ci_windows/672/

**Comment by tfoote** [comment]
That's what we talked about during the 

Is it not possible to overload the existing method which would avoid the very verbose method name?

File: rclcpp/include/rclcpp/client.hpp [hunk 1]
```diff
@@ -142,10 +163,39 @@ class Client : public ClientBase
     return f;
   }
 
+  SharedFutureWithRequest async_send_request_return_request(SharedRequest request)
```
Comment (tfoote) [inline rclcpp/include/rclcpp/client.hpp]: Do we need this method? A noop callback doesn't need the request data.

**Comment by esteve** [comment]
New CI jobs:

http://ci.ros2.org/job/ros2_batch_ci_linux/605/
http://ci.ros2.org/job/ros2_batch_ci_osx/491/
http://ci.ros2.org/job/ros2_batch_ci_windows/673/

**Comment by esteve** [comment]
I ended up using `function_traits` to fix a compilation bug on Windows because of the recurring issue with `std::function`. I've also added specializations for references and universal references to `function_traits` so that it can be used with lambdas that are not wrapped in `std::function` objects.

New CI:

http://ci.ros2.org/job/ros2_batch_ci_linux/605/
http://ci.ros2.org/job/ros2_batch_ci_osx/492/
http://ci.ros2.org/job/ros2_batch_ci_windows_opensplice/674/

**Comment by tfoote** [comment]
+1 

The connext_dynamic job failed on the filename length. @wjwwood is that fix still pending?

**Comment by esteve** [comment]
I pushed one last change to move the `function_traits` templates to a separate namespace. I could open a separate PR for it. In any case, I triggered new CI jobs:

http://ci.ros2.org/job/ros2_batch_ci_linux/607/
http://ci.ros2.org/job/ros2_batch_ci_osx/493/
http://ci.ros2.org/job/ros2_batch_ci_windows_connext_dynamic/3/
http://ci.ros2.org/job/ros2_batch_ci_windows_connext_static/3/
http://ci.ros2.org/job/ros2_batch_ci_windows_opensplice/675/

**Comment by tfoote** [comment]
+1 the new name is much clearer. There probably will be usages in other repos of the old names?

**Comment by esteve** [comment]
`function_traits` is only used in `rclcpp` itself. I updated the test in `system_tests` to use `async_send_request` instead of the longer name.

**Comment by esteve** [comment]
I also updated the code to pass uncrustify.

**Comment by dirk-thomas** [comment]
http://ci.ros2.org/job/ros2_batch_ci_linux/611/testReport/junit/rclcpp.uncrustify/xunit/include_rclcpp_function_traits_hpp/

**Comment by esteve** [comment]
https://github.com/ros2/rclcpp/commit/03697d149a8f89337c2da3acd385c0945547f96c

**Pull request merged**